pkg/controller/shoot/shoot_maintenance_control.go

```diff
@@ -131,110 +186,147 @@ func (c *defaultMaintenanceControl) Maintain(shootObj *gardenv1beta1.Shoot, key
-			case gardenv1beta1.CloudProviderOpenStack:
-				image := machineImage.(*gardenv1beta1.OpenStackMachineImage)
-				shoot.Spec.Cloud.OpenStack.MachineImage = image
-			}
-		}
-
-		// Check if the CloudProfile contains a newer Kubernetes patch version.
-		if shoot.Spec.Maintenance.AutoUpdate.KubernetesVersion {
-			newerPatchVersionFound, latestPatchVersion, err := helper.DetermineLatestKubernetesVersion(*operation.Shoot.CloudProfile, operation.Shoot.Info.Spec.Kubernetes.Version)
-			if err != nil {
-				handleError(fmt.Sprintf("Failure while determining the latest Kubernetes patch version in the CloudProfile: %s", err.Error()))
-				return nil
-			}
-			if newerPatchVersionFound {
-				shoot.Spec.Kubernetes.Version = latestPatchVersion
-			}
+		if newerPatchVersionFound {
+			shoot.Spec.Kubernetes.Version = latestPatchVersion
 		}
+	}
 
-		// Remove operation annotation.
-		delete(shoot.Annotations, common.ShootOperation)
+	// Remove operation annotation.
+	delete(shoot.Annotations, common.ShootOperation)
 
-		// Update the Shoot resource object.
-		if _, err := c.updater.UpdateShoot(shoot); err != nil {
-			handleError(fmt.Sprintf("Could not update the Shoot specification: %s", err.Error()))
-			return nil
-		}
-		msg := "Completed; updated the Shoot specification successfully."
-		shootLogger.Infof("[SHOOT MAINTENANCE] %s", msg)
-		c.recorder.Eventf(shoot, corev1.EventTypeNormal, gardenv1beta1.ShootEventMaintenanceDone, "[%s] %s", operationID, msg)
+	// Update the Shoot resource object.
+	if _, err := c.updater.UpdateShoot(shoot); err != nil {
+		handleError(fmt.Sprintf("Could not update the Shoot specification: %s", err.Error()))
+		return nil
 	}
+	msg := "Completed; updated the Shoot specification successfully."
+	shootLogger.Infof("[SHOOT MAINTENANCE] %s", msg)
+	c.recorder.Eventf(shoot, corev1.EventTypeNormal, gardenv1beta1.ShootEventMaintenanceDone, "[%s] %s", operationID, msg)
 
 	return nil
 }
 
-// NowWithinTimeWindow returns true in case the current time is within <begin> and <end>.
+const (
+	// AverageMaintenanceDuration is the average duration of a maintenance operation. It is subtracted from the end
+	// of a maintenance time window to use a best-effort kind of finishing the operation before the end.
+	// Generally, we can't make sure that the maintenance operation is done by the end of the time window anyway (considering large
+	// clusters with hundreds of nodes, a rolling update will take several hours).
+	AverageMaintenanceDuration = 15 * time.Minute
+)
+
+// TimeWithinWindow returns true in case the given time is within <begin> and <end>.
 // <begin> and <end> must be in the following format: HHMMSS+ZZZZ. In case one of the
 // times can not be parsed, an error is returned.
-func NowWithinTimeWindow(begin, end string, nowTime time.Time) (bool, error) {
-	maintenanceWindowBegin, err := utils.ParseMaintenanceTime(begin)
+func TimeWithinWindow(begin, end string, checkedTime time.Time) (bool, error) {
```

Comment:
Instead of writing functions that operate on the `begin` and `end` strings, how about introducing a type `TimeRange` / `TimeWindow` that 'owns' these methods? Once you have this `TimeWindow`, you can remove most of the `error` additional return values since they mostly result from failed parsing attempts / invalid `TimeWindows` plus the execution becomes faster.

Reply:
Sounds good, thanks. Will do.